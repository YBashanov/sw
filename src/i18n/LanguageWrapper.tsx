import React, { createContext, useState } from 'react';
import { getIntlLocale, getIntlMessages } from '@/i18n';
import { IntlProvider } from 'react-intl';

export type Language = 'ru' | 'en';

export const defaultLanguage: Language = 'ru';

type LanguageContextType = {
    language?: Language;
    setLanguage?: (language: Language) => void;
};

type LanguageWrapperType = {
    children: React.ReactNode;
};

export const LanguageContext = createContext<LanguageContextType>({
    language: defaultLanguage,
});

export const defaultSaveLang = localStorage.getItem('language') as Language;

const LanguageWrapper: React.FC<LanguageWrapperType> = ({ children }) => {
    const [language, setLanguage] = useState<Language>(defaultSaveLang || defaultLanguage);

    const handleSetLanguage = (lang: Language) => {
        localStorage.setItem('language', lang);

        // установка для React-компонентов
        setLanguage(lang);

        // установка для статики (схемы Yup, модели Redux)
        // устанавливается в i18n/index

        // требуется перезагрузка для сохранения в localStorage и взятия оттуда
        location.href = '/';
    };

    // тестирование смены языка. (в идеале это должно быть по кнопке)
    React.useEffect(() => {
        // const wantLang = 'ru';
        // if (defaultSaveLang !== wantLang) handleSetLanguage(wantLang);
    }, []);

    return (
        <LanguageContext.Provider value={{ language, setLanguage: handleSetLanguage }}>
            <IntlProvider locale={getIntlLocale(language)} messages={getIntlMessages(language)}>
                {children}
            </IntlProvider>
        </LanguageContext.Provider>
    );
};

export default LanguageWrapper;
