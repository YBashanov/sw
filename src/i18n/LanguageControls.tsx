import React, { useCallback, useContext, useMemo } from 'react';
import { useIntl } from 'react-intl';
import { Button } from 'antd';
import { Language, LanguageContext } from '@/i18n/LanguageWrapper';

const LanguageControls: React.FC = () => {
    const { language, setLanguage } = useContext(LanguageContext);
    const { formatMessage: f } = useIntl();

    const handleCheckLanguage = useCallback(
        function (lang: Language) {
            setLanguage(lang);
        },
        [setLanguage]
    );

    const buttonsData = useMemo(
        () => [
            {
                language: 'ru',
                label: 'RU',
            },
            {
                language: 'en',
                label: 'EN',
            },
        ],
        []
    );

    return (
        <div className="language-controls">
            <div className="label">{f({ id: 'App.Language.ChooseLanguage' })}</div>
            <div className="controls">
                {buttonsData.map(button => {
                    const isSelected = language === button.language;

                    return (
                        <Button
                            id={button.language}
                            key={button.language}
                            type={isSelected ? 'primary' : 'default'}
                            onClick={handleCheckLanguage.bind(this, button.language)}
                        >
                            {button.label}
                        </Button>
                    );
                })}
            </div>
        </div>
    );
};

export default LanguageControls;
