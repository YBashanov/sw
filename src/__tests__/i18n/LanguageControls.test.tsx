import React from 'react';
import { describe, expect, it } from '@jest/globals';
import { render } from 'enzyme';
import LanguageControls from '@/i18n/LanguageControls';
import LanguageWrapper from '@/i18n/LanguageWrapper';

describe('testing: Separator', function () {
    const JSX = (
        <LanguageWrapper>
            <LanguageControls />
        </LanguageWrapper>
    );
    const Component = render(JSX);

    it('Snapshot correctly', function () {
        expect(Component).toMatchSnapshot();
    });

    it('Static HTML', function () {
        expect(Component.text()).toEqual('Выбор языкаRUEN');
    });

    it('Find all buttons', () => {
        expect(Component.find('button')).toHaveLength(2);
    });
});
