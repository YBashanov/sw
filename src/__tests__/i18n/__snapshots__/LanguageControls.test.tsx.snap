// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing: Separator Snapshot correctly 1`] = `
LoadedCheerio {
  "0": <div
    class="language-controls"
  >
    <div
      class="label"
    >
      Выбор языка
    </div>
    <div
      class="controls"
    >
      <button
        class="ant-btn css-dev-only-do-not-override-ed5zg0 ant-btn-primary"
        id="ru"
        type="button"
      >
        <span>
          RU
        </span>
      </button>
      <button
        class="ant-btn css-dev-only-do-not-override-ed5zg0 ant-btn-default"
        id="en"
        type="button"
      >
        <span>
          EN
        </span>
      </button>
    </div>
  </div>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
